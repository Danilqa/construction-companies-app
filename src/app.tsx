import * as React from 'react';
import { ChangeEvent, SyntheticEvent, useCallback, useState } from 'react';
import { useFetch } from './hooks/use-fetch';
import { useSet } from './hooks/use-set';
import { Alert, Card, Col, Container, Form, Row } from 'react-bootstrap';
import { Company } from './model/company';
import { Speciality } from './model/speciality';
import { env } from './env/env';

export const App = (): JSX.Element => {
    const DELAY_FOR_SEARCH = 500;

    const [query, setQuery] = useState('');
    const {
        value: filters,
        add: setFilter,
        remove: removeFilter,
        has: isFilterSelected
    } = useSet<string>([]);

    const { responseJson: companies, isLoading, error } = useFetch<Company[]>(
        `${env.api.base}${env.api.companies}`,
        { query, specialities: filters },
        [query, filters],
        DELAY_FOR_SEARCH
    );

    const {
        responseJson: specialities,
        isLoading: isSpecialitiesLoading,
        error: specialitiesError
    } = useFetch<Speciality[]>(
        `${env.api.base}${env.api.specialities}`,
        {},
        []
    );

    const onQueryChanged = useCallback((event: ChangeEvent<HTMLInputElement>) => {
        const { value } = event.currentTarget;
        setQuery(value);
    }, []);

    const onFilterChanged = useCallback((event: SyntheticEvent<HTMLInputElement>) => {
        const { checked, name } = event.currentTarget;
        if (checked) {
            setFilter(name);
        } else {
            removeFilter(name);
        }
    }, []);

    return (
        <>
            <Container fluid>
                <Row>
                    <Col>
                        <br/>
                        <Form.Control
                            placeholder='Search'
                            onChange={onQueryChanged}
                        />
                    </Col>
                </Row>
                <br/>
                <Row>
                    <Col md={8}>
                        {error && <Alert variant='error'>{error}</Alert>}
                        {specialitiesError && <Alert variant='error'>{specialitiesError}</Alert>}
                        {!isLoading && !companies?.length && 'No companies found. Please, try a different query'}
                        {(
                            <Row>
                                {companies?.map((company) => (
                                    <Col md={3} key={company.id}>
                                        <Card>
                                            <Card.Img variant='top' src={company.logoImageUrl}/>
                                            <Card.Body>
                                                <Card.Title>{company.name}</Card.Title>
                                                <Card.Text>{company.specialties.join(', ')}</Card.Text>
                                            </Card.Body>
                                        </Card>
                                        <br/>
                                    </Col>
                                ))}
                            </Row>
                        )}
                    </Col>
                    <Col md={4}>
                        <Card>
                            <Card.Header>Speciality</Card.Header>
                            <Card.Body>
                                {isSpecialitiesLoading && 'loading...'}
                                {!isSpecialitiesLoading && specialities?.map(({ key, name }) => (
                                    <Form.Check
                                        key={key}
                                        label={name}
                                        name={key}
                                        type='checkbox'
                                        checked={isFilterSelected(key)}
                                        onChange={onFilterChanged}
                                    />
                                ))}
                            </Card.Body>
                        </Card>
                    </Col>
                </Row>
            </Container>
        </>
    );
};
