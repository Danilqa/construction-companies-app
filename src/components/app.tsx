import * as React from 'react';
import { GlobalStyles } from '../styles/global.styles';
import { Page } from './page/page';
import { memo } from 'react';
import { BrowserRouter, Routes, Route, Link } from 'react-router-dom';
import { CaseA } from '../content/case-a';
import { Home } from '../content/home';
import { CaseB } from '../content/case-b';
import { Avatar } from './avatar/avatar';
import { Header } from './header/header';

enum RouteType {
    CASE_A = 'cases/lol',
    CASE_B = 'cases/kek'
}

export const App = memo(() => {
    return (
        <React.Fragment>
            <GlobalStyles/>
            <Page.Container>
                <BrowserRouter>
                    <Page.Sidebar>
                        <Avatar/>
                        <Header>
                            S V E T I L N I C<br/>
                            W I T H - L O V E
                        </Header>

                        <Link to='/'>Home</Link><br/>
                        <Link to={RouteType.CASE_A}>Case LOL</Link><br/>
                        <Link to={RouteType.CASE_B}>Case KEK</Link>
                    </Page.Sidebar>
                    <Page.Content>
                        <Routes>
                            <Route path='/'>
                                <Route index element={<Home/>}/>
                                <Route path={RouteType.CASE_A} element={<CaseA/>}/>
                                <Route path={RouteType.CASE_B} element={<CaseB/>}/>
                            </Route>
                        </Routes>
                    </Page.Content>
                </BrowserRouter>
            </Page.Container>
        </React.Fragment>
    );
});
