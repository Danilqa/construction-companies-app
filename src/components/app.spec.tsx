import * as React from 'react';
import App from './app';
import { MemoryRouter } from 'react-router-dom';
import { Routes } from './app.enum';

describe('App', () => {

    it('should render about', async () => {
        const component = (
            <div>
                <MemoryRouter initialEntries={[Routes.ABOUT]} initialIndex={0}>
                    <App/>
                </MemoryRouter>
            </div>
        );
        await expect({ component }).toMatchComponentImageAsync();
    });

    it('should render home', async () => {
        const component = (
            <div>
                <MemoryRouter initialEntries={[Routes.HOME]} initialIndex={0}>
                    <App/>
                </MemoryRouter>
            </div>
        );
        await expect({ component }).toMatchComponentImageAsync();
    });

    it('should render hero with index 1', async () => {
        const component = (
            <div>
                <MemoryRouter initialEntries={[`${Routes.HEROES}/1`]} initialIndex={0}>
                    <App/>
                </MemoryRouter>
            </div>
        );
        await expect({ component }).toMatchComponentImageAsync();
    });
});