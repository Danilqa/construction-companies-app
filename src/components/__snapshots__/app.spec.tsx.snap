// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should be correct rendered app 1`] = `"qmt2axAqILcIxnEhDg7kHBTHyNA="`;
